### PR TITLE
Port PrusaSlicer's Measure & Cut gizmos

### LINKED ISSUE #717: [FEATURE REQUEST] Measure tool
Is there any chance of having this measure tool implemented on OS??

It allows you to select a vertex, an edge, radios or a whole face, making it easier to use the slicer itself to make precise boolean operations without having to use a 3D tool such as FreeCAD or Blender.

image
image
image
image

Thank you

### LINKED ISSUE #1150: Ruler
When importing a model, sometimes it could be very nice to have a ruler of sorts to measure the distance between two points. Could this be added?

### LINKED ISSUE #1590: Feature: viewport zoom % and/or measuring tool
**Is your feature request related to a problem? Please describe.**

You always have to be guessing dimensions scale when working in the slicer, or at best be selecting certain objects and the scale tool to have an idea of real dimensions (this works for outer dimensions only).
If you "zoom in" the view to a scale that matches your display size/ppi to real life, you are actually able to see and even measure these dimensions manually, but this means every time being able to get to an specific zoom level that you'll never be able to replicate without trial and error using some reference.
This is particularly important when you're assembling, scaling parts with inner dimensions you care about, etc.

**Describe the solution you'd like**

Two ideas come to mind. 1 seems easier and more straightforward to implement. 2 is nicer but more advanced.
1. Be able to visualize the % zoom you are at in the view pane, and ideally be able to set this % manually, even better if able to set a "default".
2. A "measure" tool to click two points/edges/faces and get a measurement, akin to what's available in other 3D tools.

**Describe how it would work**

1. Viewport zoom setting:
1.1. Example from Photoshop implementation: https://www.photoshopessentials.com/basics/photoshop-zoom/
1.2.  Example from Office implementation: https://www.excel-easy.com/examples/zoom.html
1.3. Ideally you can also save the % setting to open every time (or 1:1, fit, etc) 

2. Measurement tool examples:
2.1. https://support.shapr3d.com/hc/en-us/articles/7874465678236-Measurements
2.2. https://help.autodesk.com/view/fusion360/ENU/?guid=SLD-MEASURE

**Describe alternatives you've considered**

Right now I use a calibration cube and a transparent overlay I printed to put over the display and manually zoom in/out via trial and error to match these, then "knowing" the viewport is at "real life" zoom scale and use manual measurement tools like squares over the display to measure things.

**Additional context**
Edit: Seems a measurement tool has already been implemented in Prusaslicer 2.6.0, so maybe just bring that over: https://github.com/prusa3d/PrusaSlicer/releases/tag/version_2.6.0-alpha2

### PR DESCRIPTION
**Pull request by Noisyfox**
Build on top of #2520 and replaces it.

This PR includes:
- Rewrite the opengl rendering code, which now renders (almost) everything using shaders instead of legacy opengl function calls
- Rewriting the gizmo mouse handling code that moves the mouse handling coding into each gizmo themselves
- Rewriting the mouse picking code, now it uses ray casting to figure out what's under the mouse cursor
- Porting of the PrusaSlicer's measure tool
- Replacing existing cuting tool with the better one PrusaSlicer has
- Updating of other gizmos using PrusaSlicer's latest code base

There was a plan to also port PrusaSlicer's emboss & svg tools, but this PR is already very big and the changes needed for emboss will be even bigger and might take forever to finish. So I decided to separate them so we can get something out and start rolling out testing builds for people to play with as soon as possible.

This was developed mainly using Windows, be prepared it could have graphic issue under Linux & macOS.

Huge credit to Prusa for their amazing job!

image
image
image

Fixes #717
Fixes #1150
Fixes #1590

### CODE REVIEW

**Comment by Eldenroot** [comment]
Cool, did you use Bambustudio implementation from they github from this week? If yes, did ir help a little.bit?

**Comment by SoftFever** [comment]
> Cool, did you use Bambustudio implementation from they github from this week? If yes, did ir help a little.bit?

BambuStudio don't have such features.
These are ported from PrusaSlicer.

**Comment by Eldenroot** [comment]
Hmm, maybe I missed something, there were some commits added thus week on their repo...but I have no time to check them.one by one. I was just curious, thank you for your work!

**Comment by Noisyfox** [comment]
> Hmm, maybe I missed something, there were some commits added thus week on their repo...but I have no time to check them.one by one. I was just curious, thank you for your work!

Didn't notice that they added the snap connector until now.

**Comment by discip** [comment]
@Noisyfox
# THANK YOU! 👍🏻

**Comment by Eldenroot** [comment]
Actually, this is a huge step forward! I was a Prusa printers user (had been for over 5-6 years?) so I used only Prusa Slicer/Superslicer. Now when I switched sides, sold all Prusa machines and bought Bambulab machines... I am getting into OrcaSlicer. It is better and better with every release. Keep good work. Few more missing features from previous slicers but you are on fire, it will not take long :) Thx!!!

**Comment by netweaver1970** [comment]
Actually, I don't think of the choice between SS vs. Orca as 'siding'. Both have strengths and weaknesses. It would be a pity to have the printing world (and available time/effort) divided. SS might get some love back from his its creator https://github.com/supermerill/SuperSlicer/discussions/3902 I don't know how to politically agree on a common future for SS and Orca but I sure would love to see it somehow happen.

**Comment by igiannakas** [comment]
A few findings from some MacOS testing:

Clicking the paint tool on MacOS crashes the application. Log attached.
[Log.txt](https://github.com/SoftFever/OrcaSlicer/files/13259201/Log.txt)

Clicking on support painting on MacOS crashes the application. 
[Support.txt](https://github.com/SoftFever/OrcaSlicer/files/13259233/Support.txt)


Crashing on the seam drawing on MacOS crashes the application.
[Crash - seam.txt](https://github.com/SoftFever/OrcaSlicer/files/13259230/Crash.-.seam.txt)


UI glitch on text tool:
Screenshot 2023-11-05 at 10 47 31


Top surface colours seem a bit off - some are red, some are light red (salmon?)
Screenshot 2023-11-05 at 10 49 35


Transparency on the model when browsing layers seems to not be too transparent :) Personally I am not a fan of the model outline but that's a taste thing I suppose.
Screenshot 2023-11-05 at 10 50 48


Other than the above, I think its looks pretty OK! You also have a "debug" tool tip & ray caster pointer(?) enabled, not sure if that is intentional?

Screenshot 2023-11-05 at 10 52 54

**Comment by Noisyfox** [comment]
Thanks for your testing!



> Clicking the paint tool on MacOS crashes the application. Log attached. [Log.txt](https://github.com/SoftFever/OrcaSlicer/files/13259201/Log.txt)
> 
> Clicking on support painting on MacOS crashes the application. [Support.txt](https://github.com/SoftFever/OrcaSlicer/files/13259233/Support.txt)
> 
> Crashing on the seam drawing on MacOS crashes the application. [Crash - seam.txt](https://github.com/SoftFever/OrcaSlicer/files/13259230/Crash.-.seam.txt)

I'll take a look.



> UI glitch on text tool:
> Screenshot 2023-11-05 at 10 47 31

Are you talking about the "Embedded depth" label?



> Top surface colours seem a bit off - some are red, some are light red (salmon?) Screenshot 2023-11-05 at 10 49 35
> 
> Transparency on the model when browsing layers seems to not be too transparent :) Personally I am not a fan of the model outline but that's a taste thing I suppose. Screenshot 2023-11-05 at 10 50 48

These are caused by the changing of the rendering which makes the blending works a bit differently. I'll see if I could improve it.



> You also have a "debug" tool tip & ray caster pointer(?) enabled, not sure if that is intentional?

Ah yeah that's for debug purpose and will be disabled before the merge.

**Comment by igiannakas** [comment]
Yeap, the embedded depth label :)

**Comment by discip** [comment]
Good evening,
first of all: THANK YOU!

Here's an issue I found with this PR:
While I'm not sure if it's Linux (Manjaro) or 4K, the toolbar is a bit glitchy as you can see:
image

Since the "Embedded depth" label was mentioned here, which by the way is not attributable to this PR, I'm throwing in the following one as well:
image

thanks in advance
kind regards

**Comment by R4toLoko** [comment]
@Noisyfox everything seems to be working fine with a major issue on Linux Mint 21.1 Cinnamon ( Ubuntu 22.04 equivalent )
I use Ctrl + N like a lot with many different softwares. 
With your release, it gives me a blank screen and then it closes itself/crash.

image

The current 1.8.0-beta I am using which I don't remember which one, I can hit Ctrl + N, receive the warning about the changes

image

Discard everything and as expected, it is ready for another project

image

The measurement tool seems to be working fine tho, I never noticed PrusaSlicer actually shows the angle and that is beautiful.

**Comment by Noisyfox** [comment]
> Clicking the paint tool on MacOS crashes the application. Log attached. [Log.txt](https://github.com/SoftFever/OrcaSlicer/files/13259201/Log.txt)
> 
> Clicking on support painting on MacOS crashes the application. [Support.txt](https://github.com/SoftFever/OrcaSlicer/files/13259233/Support.txt)
> 
> Crashing on the seam drawing on MacOS crashes the application. [Crash - seam.txt](https://github.com/SoftFever/OrcaSlicer/files/13259230/Crash.-.seam.txt)

These crashes should have been fixed now.

**Comment by Noisyfox** [comment]
> While I'm not sure if it's Linux (Manjaro) or 4K, the toolbar is a bit glitchy as you can see:

Hmmmmm unfortunately I don't have a 4k monitor... I'll try if I could reproduce with my existing devices.

**Comment by Noisyfox** [comment]
> everything seems to be working fine with a major issue on Linux Mint 21.1 Cinnamon ( Ubuntu 22.04 equivalent )

I'm not able to reproduce it with my Ubuntu 22.04 VM. Do you have any crash log?

**Comment by R4toLoko** [comment]
> > everything seems to be working fine with a major issue on Linux Mint 21.1 Cinnamon ( Ubuntu 22.04 equivalent )
> 
> I'm not able to reproduce it with my Ubuntu 22.04 VM. Do you have any crash log?

Can do tomorrow afternoon, well, today afternoon ( 00:36 AM now )
Thank you.

**Comment by igiannakas** [comment]
> > Clicking the paint tool on MacOS crashes the application. Log attached. [Log.txt](https://github.com/SoftFever/OrcaSlicer/files/13259201/Log.txt)
> > Clicking on support painting on MacOS crashes the application. [Support.txt](https://github.com/SoftFever/OrcaSlicer/files/13259233/Support.txt)
> > Crashing on the seam drawing on MacOS crashes the application. [Crash - seam.txt](https://github.com/SoftFever/OrcaSlicer/files/13259230/Crash.-.seam.txt)
> 
> These crashes should have been fixed now.

Gizmos work well now. Also the alpha channel seems updated (?) and looking better!

Final thing -> the resolution of the toolbar is a bit low on a Mac as well but I think you know this already? 

image

**Comment by R4toLoko** [comment]
@Noisyfox ,

I found the problem.

From a fresh start, those layout mess is normal but that last line, isn't normal. 
It only shows up if I close the slicer which I didn't, the slicer is doing that on its own.
image

The menu resolution looks normal
image

My current 1.8.0-beta after slicing a cube and hitting Ctrl + N, no issues.

image

On the left is how your version looks, dark mode always sucked on this slicer thanks to BambuStudio, that is how it used to look for a loooong time blinding you.
On the right is my current 1.8.0-beta version with the right dark mode applied ( about time).
Something changed regarding the windowing bit on Linux and that is causing the problem, the dark mode alone is wrong so that needs to be fixed anyway.
Back to BambuLab Git ticket 12, dark mode/windowing on Linux was giving hell to everybody so I do believe this is the root cause of the problem.

image

Thank you for doing the lord works.

**Comment by Noisyfox** [comment]
> @Noisyfox ,
> 
> I found the problem.
> 
> From a fresh start, those layout mess is normal but that last line, isn't normal. It only shows up if I close the slicer which I didn't, the slicer is doing that on its own. image
> 
> The menu resolution looks normal image
> 
> My current 1.8.0-beta after slicing a cube and hitting Ctrl + N, no issues.
> 
> image
> 
> On the left is how your version looks, dark mode always sucked on this slicer thanks to BambuStudio, that is how it used to look for a loooong time blinding you. On the right is my current 1.8.0-beta version with the right dark mode applied ( about time). Something changed regarding the windowing bit on Linux and that is causing the problem, the dark mode alone is wrong so that needs to be fixed anyway. Back to BambuLab Git ticket 12, dark mode/windowing on Linux was giving hell to everybody so I do believe this is the root cause of the problem.
> 
> image
> 
> Thank you for doing the lord works.

image

Are you able to find that core dump and send it to me?

And thanks for your test, I'll see if I can improve the canvas background in dark mode.

**Comment by R4toLoko** [comment]
> Update: the dark mode background should now been fixed

I can look around but I do believe that you fixed it. Will give it another try once the action is completed.
Thanks man

**Comment by R4toLoko** [comment]
No juice.
Dark theme is fixed but the crashing continues.
I even tried to compile the slicer on my PC as I used to do, neither `master` of `SoftFever` branches compile. 
Usually, when I cannot compile the slicer on my PC, something is off, unless those branches aren't ready and I need to fetch a 1.7.0 branch or something ( sorry I am not a developer and hate coding )

Also, the file `GLCanvas3D.cpp` seems to be a lot different from master.
I am totally noob when it comes to windowing on Linux, I will see if I can find out what is wrong.

Peek 2023-11-09 09-33


[debug_Thu_Nov_09_09_28_13_10234.log](https://github.com/SoftFever/OrcaSlicer/files/13301893/debug_Thu_Nov_09_09_28_13_10234.log)


I don't know if it has anything to do with this little guy
image

**Comment by Noisyfox** [comment]
> Final thing -> the resolution of the toolbar is a bit low on a Mac as well but I think you know this already?

Yeah I noticed that overall anti-aliasing is worse after switching to shaders. Not a 3D expert and this is not a blocker to get this PR merged, but I'll try my best to see if I could fix this problem.

**Comment by igiannakas** [comment]
> > Final thing -> the resolution of the toolbar is a bit low on a Mac as well but I think you know this already?
> 
> Yeah I noticed that overall anti-aliasing is worse after switching to shaders. Not a 3D expert and this is not a blocker to get this PR merged, but I'll try my best to see if I could fix this problem.

I don’t think this was an issue in the previous version I tested above, if that helps…

**Comment by Noisyfox** [comment]
> No juice. Dark theme is fixed but the crashing continues. I even tried to compile the slicer on my PC as I used to do, neither `master` of `SoftFever` branches compile. Usually, when I cannot compile the slicer on my PC, something is off, unless those branches aren't ready and I need to fetch a 1.7.0 branch or something ( sorry I am not a developer and hate coding )
> 
> Also, the file `GLCanvas3D.cpp` seems to be a lot different from master. I am totally noob when it comes to windowing on Linux, I will see if I can find out what is wrong.
> 
> Peek 2023-11-09 09-33 [ Peek 2023-11-09 09-33 ](https://user-images.githubusercontent.com/35049512/281561476-97ed6830-637d-41da-a5bd-9a5bd471b7f0.gif) [ ](https://user-images.githubusercontent.com/35049512/281561476-97ed6830-637d-41da-a5bd-9a5bd471b7f0.gif)
> 
> [debug_Thu_Nov_09_09_28_13_10234.log](https://github.com/SoftFever/OrcaSlicer/files/13301893/debug_Thu_Nov_09_09_28_13_10234.log)
> 
> I don't know if it has anything to do with this little guy image

I think the crash has been fixed now. Your gif really helps, thanks!

**Comment by R4toLoko** [comment]
@Noisyfox yup, beautiful. Out of curiosity, what was causing that?
I might start using this one now as my main one.

**Comment by Noisyfox** [comment]
> @Noisyfox yup, beautiful. Out of curiosity, what was causing that? I might start using this one now as my main one.

I have to write some extra code to support plate selection as well as mouse interaction for those 6 icons next to the bed, which I made a few mistakes. It failed to release the mouse detector when creating new projects and caused the crash.

**Comment by discip** [comment]
@Noisyfox
First of all:
## Thank you very much for being persistent in getting this done! 😃👍🏻
The toolbar looks much nicer now.
Although in my opinion it appears to be blurrier compared to the current look. 😅
Also, I encountered what igiannakas mentioned earlier on:

> Transparency on the model when browsing layers seems to not be too transparent :) Personally I am not a fan of the model outline but that's a taste thing I suppose.

Latest commit:
image
image

Current state:
image
image

**Review by SoftFever — APPROVED** [review]

**Comment by SoftFever** [comment]
It looks all good.
@Noisyfox Thanks for the fantastic work!

For any issues, let's use ticket to track them and discord to discuss

**Comment by cpasoft** [comment]
I love it... thanks @Noisyfox. I agree, you've done a fantastic work...

**Comment by SoftFever** [comment]
> @Noisyfox
> 
> 1. ## Thank you very much for being persistent in getting this done! 😃👍🏻
>    The toolbar looks much nicer now.
>    Although in my opinion it appears to be blurrier compared to the current look. 😅
> 2. Also, I encountered what igiannakas mentioned earlier on:
> 
> > Transparency on the model when browsing layers seems to not be too transparent :) Personally I am not a fan of the model outline but that's a taste thing I suppose.
> 
> Latest commit: image image
> 
> Current state: image image

Looks like the shading effect was not turn off for the transparent mode.

**Comment by Noisyfox** [comment]
Yeah I'm working on this issue right now.

**Comment by igiannakas** [comment]
Crash log attached.
[Crash Log.txt](https://github.com/SoftFever/OrcaSlicer/files/13473267/Crash.Log.txt)

File: src/slic3r/GUI/GUI_ObjectList.cpp [hunk 1]
```diff
@@ -304,6 +310,7 @@ ObjectList::ObjectList(wxWindow* parent) :
 
 ObjectList::~ObjectList()
 {
+    delete m_objects_model;
```
Comment (igiannakas) [inline src/slic3r/GUI/GUI_ObjectList.cpp:313]: This line is causing crash when closing the application. Hence probably why the original BBL code did not have the destructor invoked.

**Comment by igiannakas** [comment]
I've opened a tiny PR to remove that line but feel that this is just a workaround, unless there is something upstream that can be debugged - https://github.com/SoftFever/OrcaSlicer/pull/2904

**Pull request merged**